apmproxy/apmserver.go

```diff
@@ -131,61 +141,112 @@ func (c *Client) PostToApmServer(ctx context.Context, agentData AgentData) error
 	c.logger.Debug("Sending data chunk to APM server")
 	resp, err := c.client.Do(req)
 	if err != nil {
-		c.SetApmServerTransportState(ctx, Failing)
+		c.UpdateStatus(ctx, Failing)
 		return fmt.Errorf("failed to post to APM server: %v", err)
 	}
-
-	//Read the response body
 	defer resp.Body.Close()
-	body, err := io.ReadAll(resp.Body)
-	if err != nil {
-		c.SetApmServerTransportState(ctx, Failing)
-		return fmt.Errorf("failed to read the response body after posting to the APM server")
+
+	// On success, the server will respond with a 202 Accepted status code and no body.
+	if resp.StatusCode == http.StatusAccepted {
+		c.UpdateStatus(ctx, Healthy)
+		return nil
 	}
 
-	if resp.StatusCode == http.StatusUnauthorized {
+	// RateLimited
+	if resp.StatusCode == http.StatusTooManyRequests {
+		c.logger.Warnf("Transport has been rate limited: response status code: %d", resp.StatusCode)
+		c.UpdateStatus(ctx, RateLimited)
+		return nil
+	}
+
+	jErr := jsonResult{}
+	if err := json.NewDecoder(resp.Body).Decode(&jErr); err != nil {
+		// non critical error.
+		// Log a warning and continue.
+		c.logger.Warnf("failed to decode response body: %v", err)
+	}
+
+	// Auth errors
+	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
 		c.logger.Warnf("Authentication with the APM server failed: response status code: %d", resp.StatusCode)
-		c.logger.Debugf("APM server response body: %v", string(body))
+		for _, err := range jErr.Errors {
+			c.logger.Warnf("failed to authenticate: document %s: message: %s", err.Document, err.Message)
+		}
+		c.UpdateStatus(ctx, Failing)
+		return nil
+	}
+
+	// ClientErrors
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		c.logger.Warnf("client error: response status code: %d", resp.StatusCode)
+		for _, err := range jErr.Errors {
+			c.logger.Warnf("client error: document %s: message: %s", err.Document, err.Message)
+		}
+		c.UpdateStatus(ctx, ClientFailing)
+		return nil
+	}
+
+	// critical errors
+	if resp.StatusCode == http.StatusInternalServerError || resp.StatusCode == http.StatusServiceUnavailable {
+		c.logger.Warnf("failed to post data to APM server: response status code: %d", resp.StatusCode)
+		for _, err := range jErr.Errors {
+			c.logger.Warnf("critical error: document %s: message: %s", err.Document, err.Message)
+		}
+		c.UpdateStatus(ctx, Failing)
 		return nil
 	}
 
-	c.SetApmServerTransportState(ctx, Healthy)
-	c.logger.Debug("Transport status set to healthy")
-	c.logger.Debugf("APM server response body: %v", string(body))
-	c.logger.Debugf("APM server response status code: %v", resp.StatusCode)
+	c.logger.Warnf("unhandled status code: %d", resp.StatusCode)
 	return nil
 }
 
-// SetApmServerTransportState takes a state of the APM server transport and updates
+// IsUnhealthy returns true if the apmproxy is not healthy.
+func (c *Client) IsUnhealthy() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.Status == Failing
```

Comment:
Given the additiondal locking and that `UpdateStatus` is called on every server response, even if suceeding ([Line 151](https://github.com/elastic/apm-aws-lambda/pull/281/files#diff-646620745ab70102b8587146dbc415a5e6a40c6fe0649830f84ade4862644ab0R151)), have you thought about changing to a `RWMutex` ?

Reply:
Great point! I've switched to a `RWMutex` and updated the method to avoid locking when possible. With this approach there should be no lock contention as long as the transport is `Healthy`